### ISSUE TITLE
SSH in-browser and key authentication options exclusives?

### ISSUE THREAD

**Issue by drustan** [warp-tech/warpgate#946]
Hello!

First of all, big thanks for this amazing project; it truly is a fantastic tool :)

Now, I've encountered an issue where, if I enable both the "in-browser" and "key" authentication methods for SSH, the in-browser validation appears to malfunction. Attempting to click the button to authorize the connection results in no action.

Cheers and thanks again!

**Comment by drustan** [comment]
Yes, I did check that. When you hit enter in the shell, the shell hangs like when you did not authorized the connection.

It's actually working well when only the "in-browser" option is activated.

It would be nice to implement a feature that displays a "connection not authorized" message where the connection hasn't been validated by the user. This would clarify the connection status.

**Comment by budachst** [comment]
Can you please tell me, how you configured that? I can't find any documentation about this feature.

**Comment by drustan** [comment]
You can configure that for each user in Config / Users / youruser

![Screenshot from 2024-02-20 17-54-24](https://github.com/warp-tech/warpgate/assets/5606292/80db138a-04c6-48da-b442-5709347544fa)

**Comment by budachst** [comment]
Ahh… cool! Really nice. As for your question about whether in-browser and key authentication are exclusives, to me it seems that they're actually inclusive. If you activate either, you can't use the in-browser auth alone, but your keys must match as well. This is actually great stuff and I am so happy, that I stumbled across this.

**Comment by SheaSmith** [comment]
I think it depends what options you have setup. For example, if you select 'Any credential', then either the key OR the OOB auth can be used. Whereas if you enable both the key and OOB/in-browser auth, then both will be required.

For example, either key or OOB/in-browser would be enabled in this scenario:
![image](https://github.com/warp-tech/warpgate/assets/51303984/6dee174d-eadf-45aa-a7cc-33e728023f11)

Whereas both key and OOB/in-browser would be required in this scenario:
![image](https://github.com/warp-tech/warpgate/assets/51303984/104df0f7-ad64-49a0-ad0e-dade5560b47d)

There seems to be a separate bug where if you have 'Any credential' enabled, only SSH keys and SSO configured for a user, and a key doesn't match (and so isn't used for auth) a password is prompted instead of keyboard interactive auth, but I'll make a separate issue for that.

**Comment by theMackabu** [comment]
Ive been having this exact same issue, temporary fix is to create another account with the name sso (or similar) and have that only be able to use browser authentication

**Comment by Annihilater** [comment]
I found that when I added a public key for the admin user in the browser, the last username part of my public key was deleted when I saved it. This may cause the local public key to not match the public key saved by the warpgate admin account when I connect with the public key, so I need to enter the admin user password to log in. Please fix it as soon as possible

**Comment by Eugeny** [comment]
I've tried to reproduce the initial issue and couldn't. Could you try connecting with `ssh -v` to see what's going on with the auth process? 

Here's my setup
<img width="566" alt="Screenshot 2024-07-25 at 11 48 08" src="https://github.com/user-attachments/assets/7d3a9eee-f936-4918-958b-f2d394cfb11b">
```❯ ssh -v 'admin:qa233@127.0.0.1' -p 2222
OpenSSH_9.6p1, LibreSSL 3.3.6
...
debug1: Authentications that can continue: password,publickey,keyboard-interactive
debug1: Offering public key: /Users/eugene/.ssh/id_ed25519 ED25519 
debug1: Server accepts key: /Users/eugene/.ssh/id_ed25519 ED25519 
debug1: Authentications that can continue: keyboard-interactive
debug1: Next authentication method: keyboard-interactive
Warpgate authentication
-----------------------------------------------------------------------
Warpgate authentication: please open the following URL in your browser:
https://warpgate.com/@warpgate#/login/b2a721d1-5236-431c-b320-165301215523

Make sure you're seeing this security key: 1 9 F 0
-----------------------------------------------------------------------

(admin:qa233@127.0.0.1) Press Enter when done: 
Authenticated to 127.0.0.1 ([127.0.0.1]:2222) using "keyboard-interactive".
debug1: channel 0: new session [client-session] (inactive timeout: 0)
debug1: Entering interactive session.
```

**Comment by ulab** [comment]
I have just tried `Password & In-browser auth` and `In-browser auth` only. 

With `In-browser auth` only, the session authentication request is shown to the user in their browser, I get a security key displayed there too, but SSH already disconnected  - and didn't display anything.

```
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering RSA public key: /home/user/.ssh/id_rsa
debug1: Authentications that can continue: keyboard-interactive
debug1: Next authentication method: keyboard-interactive
debug1: Authentications that can continue: keyboard-interactive
debug1: No more authentication methods to try.
Permission denied (keyboard-interactive).
```

`Password & In-browser auth` I can authorize in browser, then type the password and then SSH just hangs. SSH does not show any security key either.

```
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering RSA public key: /home/user/.ssh/id_rsa
debug1: Authentications that can continue: password,keyboard-interactive
debug1: Next authentication method: keyboard-interactive
debug1: Authentications that can continue: password,keyboard-interactive
debug1: Next authentication method: password
wgtestuser:target@warpgate.example.com's password:'
```

I am very new to this, so I might be doing something wrong. :-)

**Comment by ulab** [comment]
With `In-browser auth` I just noticed the message in the logs "Failed to construct external URL client_ip: ::ffff:10.234.1.251error: ExternalHostUnknown". Again, I am not sure if that is related in any way.

**Comment by Eugeny** [comment]
Yes - in your case, you need to set `external_host` in the config (and `http.external_port` if you're behind a reverse proxy and the port is not the same as in `http.listen`). Warpgate needs this info to construct the auth URL that gets send to the client. The behavior when `external_host` is missing can definitely be improved though

**Comment by ulab** [comment]
Just for reference, the port is the same and the machine is not behind a reverse proxy. But it being a docker container might mess with it too. I will try to set them tomorrow and see if it works better.

**Comment by Eugeny** [comment]
Sorry if it wasn't clear - only `http.external_port` is related to being behind a proxy and is optional.

`external_host` is strictly required to run in-browser auth for SSH.

**Comment by ulab** [comment]
I am not sure if we should keep discussing it in this issue. It has become more of a support request now. Just for reference:

`external_host` has been set. I now get the `Warpgate authentication` message when connecting with SSH and the browser shows a session needs to be authorized. 

Things that happen: 

- The security key in the browser does not always display all characters.

<img width="510" height="184" alt="Image" src="https://github.com/user-attachments/assets/c0d02593-46c1-4c88-87ae-e289fb4d5e22" />

- I am guessing I should press "Authorize` now, even though the message doesn't say so ;).

After authorizing and pressing `Enter` on the console, SSH still hangs.
